broker/src/main/java/org/apache/rocketmq/broker/pop/PopConsumerService.java

```diff
@@ -0,0 +1,699 @@
+                            retryTopicV1, 0, batchSize, filter, PopConsumerRecord.RetryType.RETRY_TOPIC_V1);
+                    }
+
+                    if (brokerConfig.isEnableRetryTopicV2()) {
+                        getMessageFuture = this.getMessageAsync(getMessageFuture, clientHost, groupId,
+                            retryTopicV2, 0, batchSize, filter, PopConsumerRecord.RetryType.RETRY_TOPIC_V2);
+                    }
+                }
+            }
+
+            return getMessageFuture.thenCompose(result -> {
+                if (result.isFound() && !result.isFifo()) {
+                    if (brokerConfig.isEnablePopBufferMerge() &&
+                        popConsumerCache != null && !popConsumerCache.isCacheFull()) {
+                        this.popConsumerCache.writeRecords(result.getPopConsumerRecordList());
+                    } else {
+                        this.popConsumerStore.writeRecords(result.getPopConsumerRecordList());
+                    }
+
+                    for (int i = 0; i < result.getGetMessageResultList().size(); i++) {
+                        GetMessageResult getMessageResult = result.getGetMessageResultList().get(i);
+                        PopConsumerRecord popConsumerRecord = result.getPopConsumerRecordList().get(i);
+
+                        // If the buffer belong retries message, the message needs to be re-encoded.
+                        // The buffer should not be re-encoded when popResponseReturnActualRetryTopic
+                        // is true or the current topic is not a retry topic.
+                        boolean recode = brokerConfig.isPopResponseReturnActualRetryTopic();
+                        if (recode && popConsumerRecord.isRetry()) {
+                            result.getGetMessageResultList().set(i, this.recodeRetryMessage(
+                                getMessageResult, popConsumerRecord.getTopicId(),
+                                popConsumerRecord.getQueueId(), result.getPopTime(), invisibleTime));
+                        }
+                    }
+                }
+                return CompletableFuture.completedFuture(result);
+            }).whenComplete((result, throwable) -> {
+                if (throwable != null) {
+                    log.error("PopConsumerService popAsync get message error",
+                        throwable instanceof CompletionException ? throwable.getCause() : throwable);
+                }
+                if (result.getMessageCount() > 0) {
+                    log.debug("PopConsumerService popAsync result, found={}, groupId={}, topicId={}, queueId={}, " +
+                            "batchSize={}, invisibleTime={}, fifo={}, attemptId={}, filter={}", result.getMessageCount(),
+                        groupId, topicId, queueId, batchSize, invisibleTime, fifo, attemptId, filter);
+                }
+                consumerLockService.unlock(groupId, topicId);
```

Comment:
In some scenarios, it seems that this unlock may not be able to go? Whether to put it in final to unlock.

Reply:
I wrapped it with try-finally.